core/src/banking_stage.rs

```diff
@@ -553,57 +547,28 @@ impl BankingStage {
     ) {
         let mut rebuffered_packet_count = 0;
         let mut consumed_buffered_packets_count = 0;
-        let buffered_packet_batches_len = buffered_packet_batches.len();
+        let buffered_packets_len = buffered_packet_batches.len();
         let mut proc_start = Measure::start("consume_buffered_process");
         let mut reached_end_of_slot: Option<EndOfSlot> = None;
 
-        RetainMut::retain_mut(buffered_packet_batches, |deserialized_packet_batch| {
-            let original_unprocessed_indexes = deserialized_packet_batch
-                .unprocessed_packets
-                .keys()
-                .cloned()
-                .collect::<Vec<usize>>();
-            if let Some(end_of_slot) = &reached_end_of_slot {
-                // We've hit the end of this slot, no need to perform more processing,
-                // just filter the remaining packets for the invalid (e.g. too old) ones
-                // if the working_bank is available
-                let mut end_of_slot_filtering_time = Measure::start("end_of_slot_filtering");
-                let should_retain = if let Some(bank) = &end_of_slot.working_bank {
-                    let new_unprocessed_indexes = Self::filter_unprocessed_packets_at_end_of_slot(
-                        bank,
-                        &deserialized_packet_batch.unprocessed_packets,
-                        my_pubkey,
-                        end_of_slot.next_slot_leader,
-                        banking_stage_stats,
-                    );
-
-                    let end_of_slot_filtered_invalid_count = original_unprocessed_indexes
-                        .len()
-                        .saturating_sub(new_unprocessed_indexes.len());
-
-                    slot_metrics_tracker.increment_end_of_slot_filtered_invalid_count(
-                        end_of_slot_filtered_invalid_count as u64,
-                    );
-
-                    banking_stage_stats
-                        .end_of_slot_filtered_invalid_count
-                        .fetch_add(end_of_slot_filtered_invalid_count, Ordering::Relaxed);
+        let mut retryable_packets = MinMaxHeap::with_capacity(buffered_packet_batches.capacity());
+        std::mem::swap(
+            &mut buffered_packet_batches.packet_priority_queue,
+            &mut retryable_packets,
+        );
 
-                    deserialized_packet_batch.update_buffered_packets_with_new_unprocessed(
-                        &original_unprocessed_indexes,
-                        &new_unprocessed_indexes,
-                    )
-                } else {
-                    true
-                };
-                end_of_slot_filtering_time.stop();
-                slot_metrics_tracker
-                    .increment_end_of_slot_filtering_us(end_of_slot_filtering_time.as_us());
-                should_retain
-            } else {
+        let mut retryable_packets: MinMaxHeap<PacketPriorityQueueEntry> = retryable_packets
+            .drain_desc()
```

Comment:
+1 for `drain + swap`, it prolly have noticeable saving from `pop + insert` MinMaxHeap.